File: R/weather_nasapower.R
# File: weather_nasapower.R

# Define the function with the *exact same arguments* as the Daymet one
process_weather_nasapower <- function(shapefile, start_year, end_year, output_dir,
                                      id_col, lat_col, lon_col, n_cores, log_file) {
  
  message(sprintf("--- Starting NASA-POWER Download (Years: %d-%d) ---", start_year, end_year))
  
  # --- Handle Date Ranges ---
  start_date_str <- paste0(start_year, "-01-01")
  
  # Check if end_year is the current year. If so, use today's date.
  current_year <- lubridate::year(Sys.Date())
  if (end_year == current_year) {
    # NASA-POWER data is usually available up to a day or two ago.
    end_date_str <- as.character(Sys.Date() - 2)
    message(sprintf("End year is current year. Fetching data up to: %s", end_date_str))
  } else {
    end_date_str <- paste0(end_year, "-12-31")
  }

  cl <- parallel::makeCluster(n_cores)
  # Safety net: release the cluster even if the download errors out before the
  # explicit stopCluster() below. try() keeps it harmless on the normal path.
  on.exit(try(parallel::stopCluster(cl), silent = TRUE), add = TRUE)
  doParallel::registerDoParallel(cl)
  message(sprintf("Registered %d cores for parallel NASA-POWER download.", n_cores))
  
  # Extract coordinates and IDs robustly
  coords_list <- .extract_coords(shapefile, id_col, lat_col, lon_col)
  ids <- coords_list$ids
  lats <- coords_list$lats
  lons <- coords_list$lons
  
  # Define the DSSAT variables from NASA-POWER
  nasa_params <- c(
    "T2M_MAX",           # Max Temp (C)
    "T2M_MIN",           # Min Temp (C)
    "ALLSKY_SFC_SW_DWN", # Solar Radiation (MJ/m^2/day)
    "PRECTOTCORR",       # Precipitation (mm/day)
    "T2MDEW",            # Dewpoint Temp (C)
    "RH2M",              # Relative Humidity (%)
    "WS2M"               # Wind speed at 2 m height (m/s)
  )
  
  foreach(i = 1:nrow(shapefile), .packages = c("nasapower", "lubridate", "dplyr")) %dopar% {
    
    latitude  <- lats[i]
    longitude <- lons[i]
    point_id  <- ids[i]
    output_file <- file.path(output_dir, sprintf("%s.WTH", point_id))
    
    if (file.exists(output_file)) {
      return(NULL)
    }
    
    tryCatch({
      
      # Download NASA-POWER data
      power_data <- nasapower::get_power(
        community    = "AG",
        lonlat       = c(longitude, latitude),
        pars         = nasa_params,
        dates        = c(start_date_str, end_date_str),
        temporal_api = "DAILY"
      )
      
      # Check if data is empty
      if (nrow(power_data) == 0) {
        stop("No data returned from NASA-POWER.")
      }
      
      # Rename and format
      weather_data <- power_data %>%
        dplyr::rename(
          SRAD = ALLSKY_SFC_SW_DWN,
          TMAX = T2M_MAX,
          TMIN = T2M_MIN,
          RAIN = PRECTOTCORR,
          TDEW = T2MDEW,
          RH2M = RH2M,
          WIND = WS2M
        ) %>%
        dplyr::mutate(
          DATE = sprintf("%d%03d", YEAR, DOY)
        ) %>%
        # NASA-POWER uses -999 for missing values
        dplyr::mutate(dplyr::across(dplyr::where(is.numeric), ~ ifelse(. == -999, -99, .)))

      weather_data <- .normalize_weather_missing_values(
        weather_data,
        point_id = point_id,
        output_file = output_file,
        output_dir = output_dir,
        source_label = "NASA_POWER"
      )
      
      # Calculate TAV and AMP
      weather_data$TAVG <- (weather_data$TMAX + weather_data$TMIN) / 2
      tav <- mean(weather_data$TAVG, na.rm = TRUE)
      
      monthly_temps <- weather_data %>%
        dplyr::group_by(YEAR, MM) %>%
        dplyr::summarise(TAVG_MON = mean(TAVG, na.rm = TRUE), .groups = "drop")
      
      annual_amps <- monthly_temps %>%
        dplyr::group_by(YEAR) %>%
        dplyr::summarise(AMP_YR = max(TAVG_MON) - min(TAVG_MON), .groups = "drop")
      
      amp <- mean(annual_amps$AMP_YR, na.rm = TRUE)
      
      # Format and write the .WTH file
      wth_header <- sprintf(
        #"$WEATHER DATA: NASA-POWER (Point ID: %s)\n@ INSI      LAT     LONG  ELEV  TAV  AMP REFHT WNDHT\n  NASA %8.4f %8.4f  -99 %5.1f %5.1f   2.0   2.0\n@ DATE  SRAD  TMAX  TMIN  RAIN  TDEW  RH2M  WIND",
        "$WEATHER DATA: NASA-POWER (Point ID: %s)\n@ INSI      LAT     LONG  ELEV   TAV   AMP REFHT WNDHT\n  NASA %8.4f %8.4f   -99 %5.1f %5.1f   2.0   2.0\n@  DATE  SRAD  TMAX  TMIN  RAIN  TDEW  RH2M  WIND",
        point_id, latitude, longitude, tav, amp
      )
      
      # Guard against values that would overflow a %6.1f field (>=9999.95 or
      # <=-999.95) and shift every downstream column. Defined locally so it is
      # always visible inside the parallel worker. Almost always a corrupt source
      # value -> write the DSSAT missing indicator (-99) instead of a broken row.
      clamp_wth <- function(x) ifelse(!is.na(x) & (x >= 9999.95 | x <= -999.95), -99, x)
      weather_lines <- with(weather_data, {
        sprintf(
          "%7s%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f",
          DATE, clamp_wth(SRAD), clamp_wth(TMAX), clamp_wth(TMIN),
          clamp_wth(RAIN), clamp_wth(TDEW), clamp_wth(RH2M), clamp_wth(WIND)
        )
      })
      
      weather_lines <- gsub("-99.0", "  -99", weather_lines, fixed = TRUE)
      wth_content   <- c(wth_header, weather_lines)
      writeLines(wth_content, con = output_file)
      
    },
    error = function(e) {
      error_message <- sprintf(
        "\n--- ERROR on task %d ---\nFailed to process point ID: %s\nCoords: Lat: %.3f, Lon: %.3f\nOriginal error: %s\n",
        i, point_id, latitude, longitude, conditionMessage(e)
      )
      cat(error_message)
      write(error_message, file = log_file, append = TRUE)
    })
    
  }  # End foreach loop
  
  parallel::stopCluster(cl)
  message(sprintf("\nNASA-POWER processing complete. Check the '%s' directory.\n", output_dir))
  
}

File: python/dssatutils/weather_nasapower.py
# File: weather_nasapower.py
# Python port of weather_nasapower.R
#
# Fetches NASA POWER daily data for each grid point and writes
# DSSAT-formatted .WTH files.  Covers the globe; available from ~1981 onward.
#
# API docs: https://power.larc.nasa.gov/docs/services/api/temporal/daily/

import os
import time
import logging
from datetime import date
from concurrent.futures import ProcessPoolExecutor, as_completed

import numpy as np
import pandas as pd
import requests

logger = logging.getLogger(__name__)

# NASA POWER parameters required for DSSAT
_NASA_PARAMS = [
    "T2M_MAX",            # Max air temperature at 2 m  (°C)
    "T2M_MIN",            # Min air temperature at 2 m  (°C)
    "ALLSKY_SFC_SW_DWN",  # Solar radiation (MJ/m²/day)
    "PRECTOTCORR",        # Precipitation (mm/day)
    "T2MDEW",             # Dew-point temperature (°C)
    "RH2M",               # Relative humidity at 2 m (%)
    "WS2M",               # Wind speed at 2 m (m/s)
]


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _calc_tav(df: pd.DataFrame) -> float:
    tavg = (df["TMAX"] + df["TMIN"]) / 2.0
    return float(tavg.mean())


def _calc_amp(df: pd.DataFrame) -> float:
    df = df.copy()
    df["TAVG"] = (df["TMAX"] + df["TMIN"]) / 2.0
    monthly = df.groupby(["YEAR", "MM"])["TAVG"].mean().reset_index()
    annual = monthly.groupby("YEAR")["TAVG"].agg(lambda x: x.max() - x.min())
    return float(annual.mean())


def _fetch_nasa_power(lat: float, lon: float, start: str, end: str,
                      retries: int = 3, backoff: float = 5.0) -> pd.DataFrame:
    """
    Fetch NASA POWER daily data for a single point.
    start / end: "YYYYMMDD" strings.
    Returns a DataFrame with columns matching _NASA_PARAMS plus YEAR, MM, DOY.
    """
    url = "https://power.larc.nasa.gov/api/temporal/daily/point"
    params = {
        "parameters": ",".join(_NASA_PARAMS),
        "community": "AG",
        "longitude": lon,
        "latitude": lat,
        "start": start,
        "end": end,
        "format": "JSON",
    }
    for attempt in range(retries):
        try:
            r = requests.get(url, params=params, timeout=180)
            r.raise_for_status()
            data = r.json()
            param_dict = data["properties"]["parameter"]
            # param_dict[param][YYYYMMDD] = value
            records = {}
            for param, daily in param_dict.items():
                records[param] = daily
            df = pd.DataFrame(records)
            df.index = pd.to_datetime(df.index, format="%Y%m%d")
            df["YEAR"] = df.index.year
            df["MM"] = df.index.month
            df["DOY"] = df.index.day_of_year
            df["DATE"] = df["YEAR"].astype(str) + df["DOY"].astype(str).str.zfill(3)
            df = df.reset_index(drop=True)
            # Replace NASA missing-value sentinel (-999) with -99
            df = df.replace(-999, -99)
            return df
        except Exception as exc:
            if attempt < retries - 1:
                time.sleep(backoff * (attempt + 1))
            else:
                raise RuntimeError(f"NASA POWER fetch failed: {exc}") from exc


def _process_single_point(args: dict) -> None:
    """Worker function executed in a subprocess."""
    lat = args["latitude"]
    lon = args["longitude"]
    pid = args["point_id"]
    output_dir = args["output_dir"]
    start_date = args["start_date"]
    end_date = args["end_date"]
    log_file = args["log_file"]

    out_path = os.path.join(output_dir, f"{pid}.WTH")
    if os.path.exists(out_path):
        return

    try:
        df = _fetch_nasa_power(lat, lon, start_date, end_date)

        if df.empty:
            raise ValueError("No data returned from NASA POWER.")

        # Rename parameter columns to DSSAT convention
        rename = {
            "ALLSKY_SFC_SW_DWN": "SRAD",
            "T2M_MAX": "TMAX",
            "T2M_MIN": "TMIN",
            "PRECTOTCORR": "RAIN",
            "T2MDEW": "TDEW",
            "RH2M": "RH2M",
            "WS2M": "WIND",
        }
        df = df.rename(columns=rename)

        tav = _calc_tav(df)
        amp = _calc_amp(df)

        # --- Write WTH file ---
        header = (
            f"$WEATHER DATA: NASA-POWER (Point ID: {pid})\n"
            f"@ INSI      LAT     LONG  ELEV   TAV   AMP REFHT WNDHT\n"
            f"  NASA {lat:8.4f} {lon:8.4f}   -99 {tav:5.1f} {amp:5.1f}   2.0   2.0\n"
            f"@  DATE  SRAD  TMAX  TMIN  RAIN  TDEW  RH2M  WIND"
        )

        lines = []
        for _, row in df.iterrows():
            line = (
                f"{row['DATE']:>7s}"
                f"{row['SRAD']:6.1f}"
                f"{row['TMAX']:6.1f}"
                f"{row['TMIN']:6.1f}"
                f"{row['RAIN']:6.1f}"
                f"{row['TDEW']:6.1f}"
                f"{row['RH2M']:6.1f}"
                f"{row['WIND']:6.1f}"
            )
            line = line.replace(" -99.0", "   -99")
            lines.append(line)

        with open(out_path, "w") as fh:
            fh.write(header + "\n")
            fh.write("\n".join(lines) + "\n")

    except Exception as exc:
        msg = (
            f"\n--- ERROR ---\n"
            f"Failed: Point ID {pid} | Lat {lat:.3f}, Lon {lon:.3f}\n"
            f"Error: {exc}\n"
        )
        print(msg)
        with open(log_file, "a") as lf:
            lf.write(msg)


# ---------------------------------------------------------------------------
# Public entry point (mirrors R function signature exactly)
# ---------------------------------------------------------------------------

def process_weather_nasapower(
    shapefile,           # GeoDataFrame
    start_year: int,
    end_year: int,
    output_dir: str,
    id_col: str,
    lat_col: str,
    lon_col: str,
    n_cores: int,
    log_file: str,
) -> None:
    """
    Download NASA POWER weather data for every point in *shapefile* and write
    DSSAT .WTH files to *output_dir*.  Mirrors the R ``process_weather_nasapower``
    function, including the current-year partial-data handling.
    """
    today = date.today()
    current_year = today.year
    start_date = f"{start_year}0101"
    if end_year == current_year:
        # NASA POWER data typically lags by a day or two
        safe_end = today - __import__("datetime").timedelta(days=2)
        end_date = safe_end.strftime("%Y%m%d")
        print(f"End year is current year. Fetching data up to: {safe_end.isoformat()}")
    else:
        end_date = f"{end_year}1231"

    print(f"--- Starting NASA-POWER Download (Years: {start_year}–{end_year}) ---")
    print(f"Registered {n_cores} cores for parallel NASA-POWER download.")

    os.makedirs(output_dir, exist_ok=True)

    tasks = []
    for _, row in shapefile.iterrows():
        tasks.append(
            dict(
                latitude=float(row[lat_col]),
                longitude=float(row[lon_col]),
                point_id=str(row[id_col]),
                output_dir=output_dir,
                start_date=start_date,
                end_date=end_date,
                log_file=log_file,
            )
        )

    with ProcessPoolExecutor(max_workers=n_cores) as pool:
        futures = {pool.submit(_process_single_point, t): t["point_id"] for t in tasks}
        for fut in as_completed(futures):
            pid = futures[fut]
            try:
                fut.result()
            except Exception as exc:
                print(f"ERROR (point {pid}): {exc}")

    print(f"\nNASA-POWER processing complete. Check the '{output_dir}' directory.\n")
